File: library/src/main/java/com/github/porokoro/paperboy/ConfigurationSerializer.java
/*
 * Copyright (C) 2015 porokoro
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package com.github.porokoro.paperboy;

import android.support.annotation.NonNull;
import android.util.JsonReader;
import android.util.JsonWriter;
import android.util.SparseArray;

import java.io.IOException;
import java.io.StringReader;
import java.io.StringWriter;

class ConfigurationSerializer {

    @NonNull
    public static String write(@NonNull PaperboyConfiguration config) {
        StringWriter stringWriter = new StringWriter();
        JsonWriter writer = new JsonWriter(stringWriter);

        try {
            writeConfiguration(writer, config);
            stringWriter.flush();
        }
        catch (IOException e) {
            e.printStackTrace();
        }
        finally {
            try {
                writer.close();
            }
            catch (IOException e) {
                e.printStackTrace();
            }
        }

        return stringWriter.toString();
    }

    private static void writeConfiguration(@NonNull JsonWriter writer, @NonNull PaperboyConfiguration config)
            throws IOException {
        writer.beginObject()
              .name("file").value(config.getFile())
              .name("viewType").value(config.getViewType())
              .name("itemLayout").value(config.getItemLayout())
              .name("sortItems").value(config.isSortItems())
              .name("itemTypes");

        writeItemTypes(writer, config.getItemTypes());

        writer.endObject();
    }

    private static void writeItemTypes(@NonNull JsonWriter writer, @NonNull SparseArray<ItemType> itemTypes)
            throws IOException {
        writer.beginArray();

        for (int i = 0; i < itemTypes.size(); i++) {
            writer.beginObject()
                  .name("key").value(itemTypes.keyAt(i))
                  .name("value");

            writeItemType(writer, itemTypes.valueAt(i));
            writer.endObject();
        }

        writer.endArray();
    }

    private static void writeItemType(@NonNull JsonWriter writer, @NonNull ItemType itemType) throws IOException {
        writer.beginObject()
              .name("id").value(itemType.getId())
              .name("name").value(itemType.getName())
              .name("shorthand").value(itemType.getShorthand())
              .name("titleSingular").value(itemType.getTitleSingular())
              .name("titlePlural").value(itemType.getTitlePlural())
              .name("color").value(itemType.getColor())
              .name("icon").value(itemType.getIcon())
              .name("sortOrder").value(itemType.getSortOrder())
              .endObject();
    }

    @NonNull
    public static PaperboyConfiguration read(@NonNull String input) {
        StringReader stringReader = new StringReader(input);
        JsonReader reader = new JsonReader(stringReader);

        try {
            return readConfiguration(reader);
        }
        catch (IOException e) {
            e.printStackTrace();
        }
        finally {
            try {
                reader.close();
            }
            catch (IOException e) {
                e.printStackTrace();
            }
        }

        return new PaperboyConfiguration();
    }

    @NonNull
    private static PaperboyConfiguration readConfiguration(@NonNull JsonReader reader) throws IOException {
        PaperboyConfiguration config = new PaperboyConfiguration();
        reader.beginObject();

        while (reader.hasNext()) {
            switch (reader.nextName()) {
                case "file":
                    config.setFile(reader.nextString());
                    break;
                case "viewType":
                    config.setViewType(ViewTypes.fromValue(reader.nextInt()));
                    break;
                case "itemLayout":
                    config.setItemLayout(reader.nextInt());
                    break;
                case "sortItems":
                    config.setSortItems(reader.nextBoolean());
                    break;
                case "itemTypes":
                    config.setItemTypes(readItemTypes(reader));
                    break;
                default:
                    reader.skipValue();
                    break;
            }
        }

        reader.endObject();
        return config;
    }

    @NonNull
    private static SparseArray<ItemType> readItemTypes(@NonNull JsonReader reader) throws IOException {
        SparseArray<ItemType> itemTypes = new SparseArray<>();
        reader.beginArray();

        while (reader.hasNext()) {
            int key = 0;
            ItemType value = null;
            reader.beginObject();

            while (reader.hasNext()) {
                switch (reader.nextName()) {
                    case "key":
                        key = reader.nextInt();
                        break;
                    case "value":
                        value = readItemType(reader);
                        break;
                    default:
                        reader.skipValue();
                        break;
                }
            }

            itemTypes.put(key, value);
            reader.endObject();
        }

        reader.endArray();
        return itemTypes;
    }

    @NonNull
    private static ItemType readItemType(@NonNull JsonReader reader) throws IOException {
        ItemType itemType = new ItemType();
        reader.beginObject();

        while (reader.hasNext()) {
            switch (reader.nextName()) {
                case "id":
                    itemType.setId(reader.nextInt());
                    break;
                case "name":
                    itemType.setName(reader.nextString());
                    break;
                case "shorthand":
                    itemType.setShorthand(reader.nextString());
                    break;
                case "titleSingular":
                    itemType.setTitleSingular(reader.nextString());
                    break;
                case "titlePlural":
                    itemType.setTitlePlural(reader.nextString());
                    break;
                case "color":
                    itemType.setColor(reader.nextInt());
                    break;
                case "icon":
                    itemType.setIcon(reader.nextInt());
                    break;
                case "sortOrder":
                    itemType.setSortOrder(reader.nextInt());
                    break;
                default:
                    reader.skipValue();
                    break;
            }
        }

        reader.endObject();
        return itemType;
    }
}

File: paperboy/src/main/kotlin/com/github/porokoro/paperboy/ConfigurationSerializer.kt
/*
 * Copyright (C) 2015-2016 porokoro
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.porokoro.paperboy

import android.util.JsonReader
import android.util.JsonToken
import android.util.JsonWriter
import android.util.SparseArray
import com.github.porokoro.paperboy.extensions.iterator
import java.io.IOException
import java.io.StringReader
import java.io.StringWriter

internal object ConfigurationSerializer {

    fun write(config: PaperboyConfiguration): String {
        val stringWriter = StringWriter()

        try {
            JsonWriter(stringWriter).use {
                writeConfiguration(it, config)
                stringWriter.flush()
            }
        } catch(e: IOException) {
            e.printStackTrace()
        }

        return stringWriter.toString()
    }

    private fun writeConfiguration(writer: JsonWriter, config: PaperboyConfiguration) {
        writer.beginObject()
                .name("file").value(config.file)
                .name("fileRes").value(config.fileRes)
                .name("viewType").value(config.viewType)
                .name("sectionLayout").value(config.sectionLayout)
                .name("typeLayout").value(config.typeLayout)
                .name("itemLayout").value(config.itemLayout)
                .name("sortItems").value(config.sortItems)
                .name("itemTypes").writeItemTypes(config.itemTypes)
                .endObject()
    }

    private fun JsonWriter.writeItemTypes(itemTypes: SparseArray<ItemType>): JsonWriter {
        beginArray()

        for ((key, value) in itemTypes) {
            beginObject()
                    .name("key").value(key)
                    .name("value").writeItemType(value)
                    .endObject()
        }

        return endArray()
    }

    private fun JsonWriter.writeItemType(itemType: ItemType) =
            beginObject()
                    .name("id").value(itemType.id)
                    .name("name").value(itemType.name)
                    .name("shorthand").value(itemType.shorthand)
                    .name("titleSingular").value(itemType.titleSingular)
                    .name("titlePlural").value(itemType.titlePlural)
                    .name("color").value(itemType.color)
                    .name("icon").value(itemType.icon)
                    .name("sortOrder").value(itemType.sortOrder)
                    .endObject()

    fun read(input: String): PaperboyConfiguration =
            try {
                JsonReader(StringReader(input)).use {
                    readConfiguration(it)
                }
            } catch(e: IOException) {
                e.printStackTrace()
                PaperboyConfiguration()
            }

    private fun readConfiguration(reader: JsonReader): PaperboyConfiguration {
        val config = PaperboyConfiguration()
        reader.beginObject()

        while (reader.hasNext()) {
            when (reader.nextName()) {
                "file" -> config.file = reader.nextStringOrNull()
                "fileRes" -> config.fileRes = reader.nextInt()
                "viewType" -> config.viewType = ViewTypes.fromValue(reader.nextInt())
                "sectionLayout" -> config.sectionLayout = reader.nextInt()
                "typeLayout" -> config.typeLayout = reader.nextInt()
                "itemLayout" -> config.itemLayout = reader.nextInt()
                "sortItems" -> config.sortItems = reader.nextBoolean()
                "itemTypes" -> config.itemTypes = readItemTypes(reader)
                else -> reader.skipValue()
            }
        }

        reader.endObject()
        return config
    }

    private fun readItemTypes(reader: JsonReader): SparseArray<ItemType> {
        val itemTypes = SparseArray<ItemType>()
        reader.beginArray()

        while (reader.hasNext()) {
            var key = 0
            var value: ItemType? = null
            reader.beginObject()

            while (reader.hasNext()) {
                when (reader.nextName()) {
                    "key" -> key = reader.nextInt()
                    "value" -> value = readItemType(reader)
                    else -> reader.skipValue()
                }
            }

            itemTypes.put(key, value)
            reader.endObject()
        }


        reader.endArray()
        return itemTypes
    }

    private fun readItemType(reader: JsonReader): ItemType {
        val itemType = ItemType()
        reader.beginObject()

        while (reader.hasNext()) {
            when (reader.nextName()) {
                "id" -> itemType.id = reader.nextInt()
                "name" -> itemType.name = reader.nextString()
                "shorthand" -> itemType.shorthand = reader.nextString()
                "titleSingular" -> itemType.titleSingular = reader.nextString()
                "titlePlural" -> itemType.titlePlural = reader.nextString()
                "color" -> itemType.color = reader.nextInt()
                "icon" -> itemType.icon = reader.nextInt()
                "sortOrder" -> itemType.sortOrder = reader.nextInt()
                else -> reader.skipValue()
            }
        }

        reader.endObject()
        return itemType
    }

    private fun JsonReader.nextStringOrNull(): String? =
            if (peek() == JsonToken.STRING)
                nextString()
            else {
                nextNull()
                null
            }
}
